File: runtime/ts/src/wire-observer.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';
import { encodePath } from '@nightseam/duplex';
import type { Endpoint, Wire } from '@bitspark/bitwire';
import { createDispatcher } from './dispatcher.ts';
import { DuplexError } from './error.ts';
import type { Observer, ObserverEvent } from './observer.ts';
import { callWire, emitWire, registerWire } from './wire.ts';
import { wirePair } from './wire-pair.ts';

const opaque = (wire: Endpoint): Endpoint => ({ send: wire.send, receive: wire.receive, close: wire.close });
const deferred = () => {
  let resolve!: () => void;
  return { promise: new Promise<void>((done) => (resolve = done)), resolve: () => resolve() };
};

test('wire helper observations label opaque model operations without adding transport observations', async () => {
  const carrier: ObserverEvent[] = [],
    events: ObserverEvent[] = [];
  const observer: Observer = { observe: (event) => events.push(event) };
  const [left, right] = wirePair({ observer: { observe: (event) => carrier.push(event) } });
  const dispatcher = createDispatcher(opaque(right));
  try {
    const delivered = deferred(),
      ended = deferred();
    const named: Observer = {
      observe: (event) => {
        observer.observe(event);
        if (event.type === 'request.ended' && event.incoming) ended.resolve();
      },
    };
    registerWire(dispatcher, ['member.with.dot'], {
      observer: named,
      family: 'probe',
      request: (params) => params,
      event: (data) => {
        assert.equal(data, 'secret😀');
        delivered.resolve();
      },
    });
    assert.equal(
      await callWire(opaque(left), ['member.with.dot'], 'secret😀', { observer, family: 'probe' }),
      'secret😀',
    );
    await ended.promise;
    emitWire(opaque(left), ['member.with.dot'], 'secret😀', { observer, family: 'probe' });
    await delivered.promise;
    const starts = events.filter((event) => event.type === 'request.started');
    const ends = events.filter((event) => event.type === 'request.ended');
    assert.equal(starts.length, 2);
    assert.deepEqual(starts.map((event) => event.incoming).sort(), [false, true]);
    assert.equal(ends.length, 2);
    assert.ok(ends.every((event) => event.outcome === 'ok' && event.durationMs >= 0));
    assert.deepEqual(
      events.filter((event) => event.type.startsWith('event.')).map((event) => event.type),
      ['event.emitted', 'event.delivered'],
    );
    for (const event of events) {
      assert.equal('family' in event && event.family, 'probe');
      assert.equal(
        'method' in event ? event.method : 'name' in event ? event.name : undefined,
        encodePath(['member.with.dot']),
      );
      if (event.type === 'event.emitted' || event.type === 'event.delivered')
        assert.equal(event.bytes, Buffer.byteLength(JSON.stringify('secret😀')));
    }
    assert.ok(!JSON.stringify(events).includes('secret'));
    assert.deepEqual(carrier, []);
  } finally {
    dispatcher.close();
    left.close();
  }
});

test('wire helper completion distinguishes a local cancellation from a same-code refusal exactly once', async () => {
  const [left, right] = wirePair();
  const dispatcher = createDispatcher(right);
  const events: ObserverEvent[] = [],
    started = deferred(),
    ended = deferred();
  const observer: Observer = {
    observe: (event) => {
      events.push(event);
      if (event.type === 'request.ended' && event.incoming) ended.resolve();
    },
  };
  try {
    registerWire(dispatcher, ['wait'], {
      observer,
      family: 'probe',
      request: (_params, context) =>
        new Promise((resolve) => {
          context.signal.addEventListener('abort', () => resolve(null), { once: true });
          started.resolve();
        }),
    });
    const controller = new AbortController();
    let endedBeforeCancel = false;
    const selected: Wire = {
      ...opaque(left),
      send: (path, message) => {
        if (message.frame.kind === 'cancel')
          endedBeforeCancel = events.some((event) => event.type === 'request.ended' && !event.incoming);
        left.send(path, message);
      },
    };
    const pending = callWire(selected, ['wait'], null, { signal: controller.signal, observer, family: 'probe' });
    await started.promise;
    controller.abort();
    await assert.rejects(pending, { code: 'cancelled' });
    await ended.promise;
    assert.equal(endedBeforeCancel, true);
    const localEnds = events.filter((event) => event.type === 'request.ended');
    assert.equal(localEnds.length, 2);
    assert.ok(localEnds.every((event) => event.outcome === 'cancelled' && event.errorCode === 'cancelled'));
    registerWire(dispatcher, ['refuse'], {
      request: () => {
        throw new DuplexError('cancelled', 'Application refusal.');
      },
    });
    await assert.rejects(callWire(left, ['refuse'], null, { observer, family: 'probe' }), { code: 'cancelled' });
    const last = events.filter((event) => event.type === 'request.ended').at(-1)!;
    assert.equal(last.outcome, 'error');
    assert.equal(last.errorCode, 'cancelled');
  } finally {
    dispatcher.close();
    left.close();
  }
});

test('wire helper observers cannot break traffic or duplicate the root panic report', async () => {
  const transport: ObserverEvent[] = [],
    model: ObserverEvent[] = [];
  const [left, right] = wirePair({ observer: { observe: (event) => transport.push(event) } });
  const dispatcher = createDispatcher(right);
  const observer: Observer = {
    observe: (event) => {
      model.push(event);
      throw new Error('broken diagnostic');
    },
  };
  try {
    registerWire(dispatcher, ['panic'], {
      observer,
      family: 'probe',
      request: () => {
        throw new Error('application panic');
      },
    });
    await assert.rejects(callWire(left, ['panic'], null, { observer, family: 'probe' }), { code: 'internal' });
    assert.equal(transport.filter((event) => event.type === 'handler.panic').length, 1);
    assert.equal(model.filter((event) => event.type === 'handler.panic').length, 0);
    assert.equal(model.filter((event) => event.type === 'request.started').length, 2);
    assert.equal(model.filter((event) => event.type === 'request.ended').length, 2);
  } finally {
    dispatcher.close();
    left.close();
  }
});

test('wire handler refusal remains an error when cancellation raced its return', async () => {
  const [left, right] = wirePair();
  const dispatcher = createDispatcher(right);
  const events: ObserverEvent[] = [],
    started = deferred(),
    ended = deferred();
  const observer: Observer = {
    observe: (event) => {
      events.push(event);
      if (event.type === 'request.ended' && event.incoming) ended.resolve();
    },
  };
  try {
    registerWire(dispatcher, ['refuse'], {
      observer,
      request: (_params, context) =>
        new Promise((_resolve, reject) => {
          context.signal.addEventListener('abort', () => reject(new DuplexError('cancelled', 'Application refusal.')), {
            once: true,
          });
          started.resolve();
        }),
    });
    const controller = new AbortController();
    const pending = callWire(left, ['refuse'], null, { signal: controller.signal, observer });
    await started.promise;
    controller.abort();
    await assert.rejects(pending, { code: 'cancelled' });
    await ended.promise;
    const incoming = events.find((event) => event.type === 'request.ended' && event.incoming);
    assert.ok(incoming?.type === 'request.ended');
    assert.equal(incoming.outcome, 'error');
    assert.equal(incoming.errorCode, 'cancelled');
  } finally {
    dispatcher.close();
    left.close();
  }
});

test('concurrent model lifetimes have distinct observer identities without changing frame request IDs', async () => {
  const [left, right] = wirePair();
  const dispatcher = createDispatcher(right);
  const first = deferred(),
    second = deferred(),
    arrived = deferred();
  const active = new Map<string, boolean>();
  const ids: string[] = [],
    frameIds: string[] = [];
  let collision = false,
    unmatched = false;
  const observer: Observer = {
    observe: (event) => {
      if (event.type === 'request.started') {
        collision ||= active.has(event.id);
        active.set(event.id, event.incoming);
        ids.push(event.id);
      } else if (event.type === 'request.ended') {
        unmatched ||= !active.delete(event.id);
      }
    },
  };
  try {
    registerWire(dispatcher, ['hold'], {
      observer,
      request: async (value, context) => {
        frameIds.push(context.requestId);
        if (frameIds.length === 2) arrived.resolve();
        await (value === 1 ? first.promise : second.promise);
        return value;
      },
    });
    const a = callWire(left, ['hold'], 1, { observer });
    const b = callWire(left, ['hold'], 2, { observer });
    await arrived.promise;
    const count = active.size;
    second.resolve();
    assert.equal(await b, 2);
    first.resolve();
    assert.equal(await a, 1);
    assert.equal(collision, false);
    assert.equal(unmatched, false);
    assert.equal(count, 4);
    assert.equal(active.size, 0);
    assert.equal(new Set(ids).size, 4);
    assert.ok(ids.every((id) => id.startsWith('wire:')));
    assert.deepEqual(frameIds, ['c:1', 'c:1']);
  } finally {
    first.resolve();
    second.resolve();
    dispatcher.close();
    left.close();
  }
});

test('incoming model observation reports the bounded refusal selected for an unencodable response', async (t) => {
  for (const kind of ['oversized result', 'oversized public refusal', 'unencodable result']) {
    await t.test(kind, async () => {
      const [left, right] = wirePair({ maxFrameBytes: 512 });
      const dispatcher = createDispatcher(right);
      const ended = deferred();
      let incoming: ObserverEvent | undefined;
      const observer: Observer = {
        observe: (event) => {
          if (event.type === 'request.ended' && event.incoming) {
            incoming = event;
            ended.resolve();
          }
        },
      };
      try {
        registerWire(dispatcher, ['response'], {
          observer,
          family: 'probe',
          request: () => {
            if (kind === 'oversized public refusal') throw new DuplexError('denied', 'Refused.', 'x'.repeat(2048));
            return kind === 'oversized result' ? 'x'.repeat(2048) : () => {};
          },
        });
        await assert.rejects(callWire(left, ['response']), { code: 'internal' });
        await ended.promise;
        assert.ok(incoming?.type === 'request.ended');
        assert.equal(incoming.outcome, 'error');
        assert.equal(incoming.errorCode, 'internal');
      } finally {
        dispatcher.close();
        left.close();
      }
    });
  }
});

File: runtime/go/wire_observer_test.go
package runtime

import (
	"context"
	"encoding/json"
	"errors"
	bitwire "github.com/Bitspark/bitwire/wire/go"
	"strings"
	"sync"
	"testing"
	"time"

	"github.com/Bitspark/nightseam/duplex/go"
)

// Expose only the neutral contract, with no concrete peer available to inspect.
type observedOpaqueWire struct{ bitwire.Wire }
type observedOpaqueEndpoint struct{ bitwire.Endpoint }

type cancellationObservingWire struct {
	bitwire.Wire
	observer *wireObservations
	ended    chan bool
}

func (w cancellationObservingWire) Send(path []string, message bitwire.Message) error {
	if message.Frame.Kind == bitwire.ProfileCancel {
		ended := false
		for _, event := range w.observer.snapshot() {
			if event, ok := event.(RequestEnded); ok && !event.Incoming {
				ended = true
			}
		}
		w.ended <- ended
	}
	return w.Wire.Send(path, message)
}

type wireObservations struct {
	mu     sync.Mutex
	events []ObserverEvent
	ended  chan struct{}
}

func (o *wireObservations) Observe(event ObserverEvent) {
	o.mu.Lock()
	o.events = append(o.events, event)
	o.mu.Unlock()
	if end, ok := event.(RequestEnded); ok && end.Incoming && o.ended != nil {
		o.ended <- struct{}{}
	}
}
func (o *wireObservations) snapshot() []ObserverEvent {
	o.mu.Lock()
	defer o.mu.Unlock()
	return append([]ObserverEvent(nil), o.events...)
}

func TestWireHelperObservationsLabelOpaqueOperationsWithoutTransportEvents(t *testing.T) {
	carrier, model := &wireObservations{}, &wireObservations{ended: make(chan struct{}, 1)}
	a, b := localPair(t, Options{Observer: carrier})
	left, right := observedOpaqueWire{a}, observedOpaqueEndpoint{b}
	path := []string{"member.with.dot"}
	delivered := make(chan struct{})
	rightBinding := testBinding(t, right)
	_, err := RegisterWire(rightBinding, path, WireHandlers{Observer: model, Family: "probe",
		Request: func(_ context.Context, params json.RawMessage) (any, error) { return params, nil },
		Event: func(_ context.Context, data json.RawMessage) error {
			if string(data) != `"secret😀"` {
				t.Errorf("event data %s", data)
			}
			close(delivered)
			return nil
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	var result string
	if err := CallWire(context.Background(), left, path, "secret😀", &result, WireCallOptions{Observer: model, Family: "probe"}); err != nil || result != "secret😀" {
		t.Fatalf("call %q: %v", result, err)
	}
	<-model.ended
	if err := EmitWire(context.Background(), left, path, "secret😀", WireEmitOptions{Observer: model, Family: "probe"}); err != nil {
		t.Fatal(err)
	}
	<-delivered
	name, _ := duplex.EncodePath(path)
	starts, ends, emitted, received := 0, 0, 0, 0
	for _, event := range model.snapshot() {
		var family, operation string
		switch event := event.(type) {
		case RequestStarted:
			starts++
			family, operation = event.Family, event.Method
		case RequestEnded:
			ends++
			family, operation = event.Family, event.Method
			if event.Outcome != OutcomeOK || event.Duration < 0 {
				t.Errorf("outcome %#v", event)
			}
		case EventEmitted:
			emitted++
			family, operation = event.Family, event.Name
			if event.Bytes != len(`"secret😀"`) {
				t.Errorf("emitted bytes %d", event.Bytes)
			}
		case EventDelivered:
			received++
			family, operation = event.Family, event.Name
		default:
			t.Fatalf("unexpected model event %T", event)
		}
		if family != "probe" || operation != name {
			t.Errorf("event %#v", event)
		}
	}
	if starts != 2 || ends != 2 || emitted != 1 || received != 1 {
		t.Fatalf("counts %d %d %d %d", starts, ends, emitted, received)
	}
	data, _ := json.Marshal(model.snapshot())
	if strings.Contains(string(data), "secret") {
		t.Fatal("observer received payload")
	}
	if len(carrier.snapshot()) != 0 {
		t.Fatalf("unexpected transport events %#v", carrier.snapshot())
	}
}

func TestWireHelperCompletionDistinguishesCancellationFromSameCodeRefusal(t *testing.T) {
	a, b := localPair(t, Options{})
	model := &wireObservations{ended: make(chan struct{}, 1)}
	started := make(chan struct{})
	bBinding := testBinding(t, b)
	_, err := RegisterWire(bBinding, []string{"wait"}, WireHandlers{Observer: model, Family: "probe", Request: func(ctx context.Context, _ json.RawMessage) (any, error) {
		close(started)
		<-ctx.Done()
		return nil, ctx.Err()
	}})
	if err != nil {
		t.Fatal(err)
	}
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	finished := make(chan error, 1)
	selected := cancellationObservingWire{Wire: a, observer: model, ended: make(chan bool, 1)}
	go func() {
		finished <- CallWire(ctx, selected, []string{"wait"}, nil, nil, WireCallOptions{Observer: model, Family: "probe"})
	}()
	<-started
	cancel()
	if err := <-finished; !errors.Is(err, context.Canceled) {
		t.Fatalf("cancel %v", err)
	}
	<-model.ended
	if !<-selected.ended {
		t.Error("local completion was observed after cancellation admission")
	}
	ends := 0
	for _, event := range model.snapshot() {
		if event, ok := event.(RequestEnded); ok {
			ends++
			if event.Outcome != OutcomeCancelled || event.ErrorCode != "cancelled" {
				t.Errorf("end %#v", event)
			}
		}
	}
	if ends != 2 {
		t.Fatalf("ended %d times", ends)
	}
	_, err = HandleWire(bBinding, []string{"refuse"}, func(context.Context, json.RawMessage) (any, error) {
		return nil, &PublicError{Code: "cancelled", Message: "Application refusal"}
	})
	if err != nil {
		t.Fatal(err)
	}
	if err := CallWire(context.Background(), a, []string{"refuse"}, nil, nil, WireCallOptions{Observer: model, Family: "probe"}); err == nil {
		t.Fatal("missing refusal")
	}
	events := model.snapshot()
	last := events[len(events)-1].(RequestEnded)
	if last.Outcome != OutcomeErrored || last.ErrorCode != "cancelled" {
		t.Fatalf("refusal %#v", last)
	}
}

func TestWireHelperObserverCannotBreakTrafficOrDuplicatePanic(t *testing.T) {
	carrier, model := &wireObservations{}, &wireObservations{ended: make(chan struct{}, 1)}
	a, b := localPair(t, Options{Observer: carrier})
	observer := localTestObserver(func(event ObserverEvent) { model.Observe(event); panic("broken diagnostic") })
	bBinding := testBinding(t, b)
	_, err := RegisterWire(bBinding, []string{"panic"}, WireHandlers{Observer: observer, Family: "probe", Request: func(context.Context, json.RawMessage) (any, error) { panic("application panic") }})
	if err != nil {
		t.Fatal(err)
	}
	if err := CallWire(context.Background(), a, []string{"panic"}, nil, nil, WireCallOptions{Observer: observer, Family: "probe"}); err == nil {
		t.Fatal("missing panic refusal")
	}
	<-model.ended
	panics := 0
	for _, event := range carrier.snapshot() {
		if _, ok := event.(HandlerPanic); ok {
			panics++
		}
	}
	if panics != 1 {
		t.Fatalf("panic reports %d", panics)
	}
	starts, ends := 0, 0
	for _, event := range model.snapshot() {
		switch event.(type) {
		case RequestStarted:
			starts++
		case RequestEnded:
			ends++
		case HandlerPanic:
			t.Fatal("duplicate panic report")
		}
	}
	if starts != 2 || ends != 2 {
		t.Fatalf("request lifecycle %d starts %d ends", starts, ends)
	}
}

func TestWireHelperHandlerRefusalRemainsAnErrorAfterCancellation(t *testing.T) {
	a, b := localPair(t, Options{})
	model := &wireObservations{ended: make(chan struct{}, 1)}
	started := make(chan struct{})
	bBinding := testBinding(t, b)
	_, err := RegisterWire(bBinding, []string{"refuse"}, WireHandlers{Observer: model, Request: func(ctx context.Context, _ json.RawMessage) (any, error) {
		close(started)
		<-ctx.Done()
		return nil, &PublicError{Code: "cancelled", Message: "Application refusal"}
	}})
	if err != nil {
		t.Fatal(err)
	}
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	finished := make(chan error, 1)
	go func() { finished <- CallWire(ctx, a, []string{"refuse"}, nil, nil, WireCallOptions{Observer: model}) }()
	<-started
	cancel()
	if err := <-finished; !errors.Is(err, context.Canceled) {
		t.Fatalf("cancel %v", err)
	}
	<-model.ended
	for _, event := range model.snapshot() {
		if event, ok := event.(RequestEnded); ok && event.Incoming {
			if event.Outcome != OutcomeErrored || event.ErrorCode != "cancelled" {
				t.Fatalf("refusal %#v", event)
			}
			return
		}
	}
	t.Fatal("missing incoming completion")
}

func TestWireHelperConcurrentObserverIdentitiesDoNotCollide(t *testing.T) {
	a, b := localPair(t, Options{})
	var mu sync.Mutex
	active := map[string]bool{}
	ids := map[string]bool{}
	collision, unmatched := false, false
	ended := make(chan struct{}, 4)
	observer := localTestObserver(func(event ObserverEvent) {
		mu.Lock()
		defer mu.Unlock()
		switch event := event.(type) {
		case RequestStarted:
			if _, exists := active[event.ID]; exists {
				collision = true
			}
			active[event.ID] = event.Incoming
			ids[event.ID] = true
		case RequestEnded:
			if _, exists := active[event.ID]; !exists {
				unmatched = true
			}
			delete(active, event.ID)
			ended <- struct{}{}
		}
	})
	started := make(chan struct{}, 2)
	first, second := make(chan struct{}), make(chan struct{})
	bBinding := testBinding(t, b)
	_, err := RegisterWire(bBinding, []string{"hold"}, WireHandlers{Observer: observer, Request: func(_ context.Context, params json.RawMessage) (any, error) {
		started <- struct{}{}
		if string(params) == "1" {
			<-first
		} else {
			<-second
		}
		return params, nil
	}})
	if err != nil {
		t.Fatal(err)
	}
	call := func(value int) <-chan error {
		done := make(chan error, 1)
		go func() {
			var result int
			err := CallWire(context.Background(), a, []string{"hold"}, value, &result, WireCallOptions{Observer: observer})
			if err == nil && result != value {
				err = errors.New("wrong result")
			}
			done <- err
		}()
		return done
	}
	one, two := call(1), call(2)
	<-started
	<-started
	mu.Lock()
	count := len(active)
	mu.Unlock()
	close(second)
	if err := <-two; err != nil {
		t.Error(err)
	}
	close(first)
	if err := <-one; err != nil {
		t.Error(err)
	}
	// Returning the reply can wake CallWire before the incoming observation.
	deadline := time.NewTimer(5 * time.Second)
	defer deadline.Stop()
	for range 4 {
		select {
		case <-ended:
		case <-deadline.C:
			t.Fatal("missing request completion observation")
		}
	}
	mu.Lock()
	defer mu.Unlock()
	if collision || unmatched || count != 4 || len(active) != 0 || len(ids) != 4 {
		t.Fatalf("observer correlation collision=%v unmatched=%v active=%d count=%d ids=%v", collision, unmatched, len(active), count, ids)
	}
	for id := range ids {
		if !strings.HasPrefix(id, "wire:") {
			t.Errorf("model observation has physical-looking id %q", id)
		}
	}
}

func TestWireHelperObservesSelectedBoundedResponseRefusal(t *testing.T) {
	for _, kind := range []string{"oversized result", "oversized public refusal", "unencodable result"} {
		t.Run(kind, func(t *testing.T) {
			a, b := localPair(t, Options{MaxFrameBytes: 512})
			bBinding := testBinding(t, b)
			model := &wireObservations{ended: make(chan struct{}, 1)}
			_, err := RegisterWire(bBinding, []string{"response"}, WireHandlers{Observer: model, Family: "probe", Request: func(context.Context, json.RawMessage) (any, error) {
				switch kind {
				case "oversized result":
					return strings.Repeat("x", 2048), nil
				case "oversized public refusal":
					return nil, &PublicError{Code: "denied", Message: "Refused", Data: json.RawMessage(`"` + strings.Repeat("x", 2048) + `"`)}
				default:
					return func() {}, nil
				}
			}})
			if err != nil {
				t.Fatal(err)
			}
			var public *PublicError
			if err := CallWire(context.Background(), a, []string{"response"}, nil, nil); !errors.As(err, &public) || public.Code != "internal" {
				t.Fatalf("reply %v", err)
			}
			<-model.ended
			for _, event := range model.snapshot() {
				if event, ok := event.(RequestEnded); ok {
					if event.Outcome != OutcomeErrored || event.ErrorCode != "internal" {
						t.Fatalf("observed reply %#v", event)
					}
					return
				}
			}
			t.Fatal("missing completion")
		})
	}
}
